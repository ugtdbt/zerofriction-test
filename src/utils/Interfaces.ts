export interface State<T> {
  data: T;
  valideAddressForm: boolean;
  valideContactDetailsForm: boolean;
  valideOrganizationDetailsForm: boolean;
  success: boolean;
  save: boolean;
  cancel: boolean;
  formtouched: boolean;
  formreset: boolean;
}

export interface Action<T> {
  type: string;
  payload?: T;
}
