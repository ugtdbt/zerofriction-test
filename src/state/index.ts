export * as actionCreators from "./actionCreators/organisationConfigurationActionCreator";
export * from "./store";
export * from "./reducers/index";
