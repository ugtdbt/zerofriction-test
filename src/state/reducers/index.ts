import { combineReducers } from "redux";
import organisationConfigurationReducer from "./organisationConfigurationReducer";

const reducers = combineReducers({
  organisationConfiguration: organisationConfigurationReducer,
});

export default reducers;

export type State = ReturnType<typeof reducers>;
