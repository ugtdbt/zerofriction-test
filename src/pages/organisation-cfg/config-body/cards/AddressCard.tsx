import React, { memo, useState, useEffect } from "react";
import { Card, Form, InputNumber, Input } from "antd";
import type { State } from "../../../../state";
import { useSelector, useDispatch } from "react-redux";
import type { AddressType } from "../../../../utils/types/organisation-config";
import type { FormInstance } from "antd/lib/form";
import { bindActionCreators } from "redux";
import { actionCreators } from "../../../../state";

const AddressCard: React.FC = memo(() => {
  const formRef = React.useRef<FormInstance>(null);
  const initialValidationData = {
    streetName: {
      validateStatus: "",
      help: "",
    },
    streetNumber: {
      validateStatus: "",
      help: "",
    },
    postalCode: {
      validateStatus: "",
      help: "",
    },
    city: {
      validateStatus: "",
      help: "",
    },
    country: {
      validateStatus: "",
      help: "",
    },
  };
  const { data, formreset, save } = useSelector(
    (state: State) => state.organisationConfiguration
  );

  const dispatch = useDispatch();
  const { valideAddressForm, clickAddressSaveButton } = bindActionCreators(
    actionCreators,
    dispatch
  );

  useEffect(() => {
    if (save) {
      clickAddressSaveButton(formValues);
      setFormDisabled(true);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [save]);

  useEffect(() => {
    formRef.current?.resetFields();
    setFormErrors(initialValidationData);
    setFormDisabled(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [formreset]);

  const [formValues, setFormValues] = useState<AddressType>(
    data.organizationConfig.address
  );
  const [formErrors, setFormErrors] = useState(initialValidationData);
  const [formDisabled, setFormDisabled] = useState<boolean>(false);

  const handleAddressChange = (e: any) => {
    const { id, value } = e.target;
    setFormValues({ ...formValues, [id]: value });
    validateField({ ...formValues, [id]: value }, id);
    validateForm({ ...formValues, [id]: value });
  };

  const handleStreetNumberChange = (value: any) => {
    setFormValues({ ...formValues, streetNumber: parseInt(value) });
    validateField(
      { ...formValues, streetNumber: parseInt(value) },
      "streetNumber"
    );
    validateForm({ ...formValues, streetNumber: parseInt(value) });
  };

  const handlePostalCodeChange = (value: any) => {
    setFormValues({ ...formValues, postalCode: parseInt(value) });
    validateField({ ...formValues, postalCode: parseInt(value) }, "postalCode");
    validateForm({ ...formValues, postalCode: parseInt(value) });
  };

  const validateForm = (formValues: AddressType) => {
    let fromValide = true;
    if (formValues["streetName"] === "") {
      fromValide = false;
    }
    if (!formValues["streetNumber"]) {
      fromValide = false;
    }
    if (!formValues["postalCode"]) {
      fromValide = false;
    }
    if (formValues["city"] === "") {
      fromValide = false;
    }
    if (formValues["country"] === "") {
      fromValide = false;
    }
    valideAddressForm(fromValide);
  };

  const validateField = (formValues: AddressType, key: string) => {
    switch (key) {
      case "streetName":
        if (formValues[key] === "") {
          setFormErrors({
            ...formErrors,
            streetName: {
              validateStatus: "error",
              help: "streetName is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            streetName: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "streetNumber":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            streetNumber: {
              validateStatus: "error",
              help: "streetNumber is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            streetNumber: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "postalCode":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            postalCode: {
              validateStatus: "error",
              help: "postalCode is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            postalCode: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "city":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            city: {
              validateStatus: "error",
              help: "City is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            city: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      case "country":
        if (!formValues[key]) {
          setFormErrors({
            ...formErrors,
            country: {
              validateStatus: "error",
              help: "Country is required!",
            },
          });
        } else {
          setFormErrors({
            ...formErrors,
            country: {
              validateStatus: "",
              help: "",
            },
          });
        }
        break;

      default:
        break;
    }
  };

  return (
    <Card title="Address" data-testid="address-form">
      <Form
        labelCol={{ span: 10 }}
        wrapperCol={{ span: 14 }}
        layout="horizontal"
        style={{ maxWidth: 600 }}
        ref={formRef}
        disabled={formDisabled}
      >
        <Form.Item
          label="Street Name"
          name="streetName"
          validateStatus={
            formErrors.streetName.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.streetName.help !== ""
              ? formErrors.streetName.help
              : undefined
          }
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>

        <Form.Item
          label="Street Number"
          name="streetNumber"
          validateStatus={
            formErrors.streetNumber.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.streetNumber.help !== ""
              ? formErrors.streetNumber.help
              : undefined
          }
        >
          <InputNumber
            onChange={handleStreetNumberChange}
            onFocus={handleAddressChange}
            style={{ width: "100%" }}
          />
        </Form.Item>

        <Form.Item
          label="Postal Code"
          name="postalCode"
          validateStatus={
            formErrors.postalCode.validateStatus === "error"
              ? "error"
              : undefined
          }
          help={
            formErrors.postalCode.help !== ""
              ? formErrors.postalCode.help
              : undefined
          }
        >
          <InputNumber
            onChange={handlePostalCodeChange}
            onFocus={handleAddressChange}
            style={{ width: "100%" }}
          />
        </Form.Item>

        <Form.Item
          label="City"
          name="city"
          validateStatus={
            formErrors.city.validateStatus === "error" ? "error" : undefined
          }
          help={formErrors.city.help !== "" ? formErrors.city.help : undefined}
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>

        <Form.Item
          label="Country"
          name="country"
          validateStatus={
            formErrors.country.validateStatus === "error" ? "error" : undefined
          }
          help={
            formErrors.country.help !== "" ? formErrors.country.help : undefined
          }
        >
          <Input onChange={handleAddressChange} onFocus={handleAddressChange} />
        </Form.Item>
      </Form>
    </Card>
  );
});

AddressCard.displayName = "AddressCard";
export default AddressCard;
