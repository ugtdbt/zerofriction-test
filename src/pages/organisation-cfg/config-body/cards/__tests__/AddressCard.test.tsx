import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import AddressCard from "../AddressCard";
import { store } from "../../../../../state/store";
import { Provider } from "react-redux";
import "../../../../../utils/matchMedia";

describe("AddressCard component", () => {
  const AddressCardComponent = (
    <React.StrictMode>
      <Provider store={store}>
        <AddressCard />
      </Provider>
    </React.StrictMode>
  );

  afterEach(() => {
    cleanup();
  });

  it("AddressCardComponent Match Snapshot", () => {
    const tree = renderer.create(AddressCardComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render AddressCardComponent component", () => {
    render(AddressCardComponent);
    const mainElem = screen.getByTestId("address-form");
    expect(mainElem).toBeInTheDocument();
  });
});
