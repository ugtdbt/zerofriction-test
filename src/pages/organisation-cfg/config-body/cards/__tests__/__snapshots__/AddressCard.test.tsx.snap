// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressCard component AddressCardComponent Match Snapshot 1`] = `
<div
  className="ant-card ant-card-bordered css-dev-only-do-not-override-1km3mtt"
  data-testid="address-form"
>
  <div
    className="ant-card-head"
    style={Object {}}
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        Address
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <form
      className="ant-form ant-form-horizontal css-dev-only-do-not-override-1km3mtt"
      onReset={[Function]}
      onSubmit={[Function]}
      style={
        Object {
          "maxWidth": 600,
        }
      }
    >
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="streetName"
              title="Street Name"
            >
              Street Name
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="streetName"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="streetNumber"
              title="Street Number"
            >
              Street Number
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <div
                  className="ant-input-number ant-input-number-in-form-item css-dev-only-do-not-override-1km3mtt"
                  onBeforeInput={[Function]}
                  onBlur={[Function]}
                  onCompositionEnd={[Function]}
                  onCompositionStart={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                >
                  <div
                    className="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled={false}
                      aria-label="Increase Value"
                      className="ant-input-number-handler ant-input-number-handler-up"
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        className="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled={false}
                      aria-label="Decrease Value"
                      className="ant-input-number-handler ant-input-number-handler-down"
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        className="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    className="ant-input-number-input-wrap"
                  >
                    <input
                      aria-valuenow={null}
                      autoComplete="off"
                      className="ant-input-number-input"
                      disabled={false}
                      id="streetNumber"
                      onChange={[Function]}
                      onFocus={[Function]}
                      role="spinbutton"
                      step={1}
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="postalCode"
              title="Postal Code"
            >
              Postal Code
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <div
                  className="ant-input-number ant-input-number-in-form-item css-dev-only-do-not-override-1km3mtt"
                  onBeforeInput={[Function]}
                  onBlur={[Function]}
                  onCompositionEnd={[Function]}
                  onCompositionStart={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                >
                  <div
                    className="ant-input-number-handler-wrap"
                  >
                    <span
                      aria-disabled={false}
                      aria-label="Increase Value"
                      className="ant-input-number-handler ant-input-number-handler-up"
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="up"
                        className="anticon anticon-up ant-input-number-handler-up-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="up"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                    <span
                      aria-disabled={false}
                      aria-label="Decrease Value"
                      className="ant-input-number-handler ant-input-number-handler-down"
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      role="button"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        className="anticon anticon-down ant-input-number-handler-down-inner"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                  <div
                    className="ant-input-number-input-wrap"
                  >
                    <input
                      aria-valuenow={null}
                      autoComplete="off"
                      className="ant-input-number-input"
                      disabled={false}
                      id="postalCode"
                      onChange={[Function]}
                      onFocus={[Function]}
                      role="spinbutton"
                      step={1}
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="city"
              title="City"
            >
              City
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="city"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-form-item css-dev-only-do-not-override-1km3mtt"
      >
        <div
          className="ant-row ant-form-item-row css-dev-only-do-not-override-1km3mtt"
          style={Object {}}
        >
          <div
            className="ant-col ant-col-10 ant-form-item-label css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <label
              className=""
              htmlFor="country"
              title="Country"
            >
              Country
            </label>
          </div>
          <div
            className="ant-col ant-col-14 ant-form-item-control css-dev-only-do-not-override-1km3mtt"
            style={Object {}}
          >
            <div
              className="ant-form-item-control-input"
            >
              <div
                className="ant-form-item-control-input-content"
              >
                <input
                  className="ant-input css-dev-only-do-not-override-1km3mtt"
                  disabled={false}
                  id="country"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  style={Object {}}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
